feat: add catalogue per network view M10350

mswertz opened the pull request:
motivation:
* we want to run multiple catalogues on one server
* further develop networks as a starting point

result:
* defined a 'global' landing page
* enable navigation of resources scoped by a network (should become: per catalogue)
* add generic truncate component for all the 'read more' use cases
* extracted some pages in reusable components so they can be reused on different routes 

Notes:
* most changes come from adding icons
* we treat networks and organisations as nested 'catalogues' conform dcat spec.
* long term objective is to allow landing pages for each of these catalogues with their own styling

connoratrug reviewed (changes requested):
please fix conflixts

connoratrug commented on apps/nuxt3-ssr/app.vue:
```diff
@@ -9,7 +9,8 @@
     <div class="z-30 relative min-h-screen flex flex-col">
       <slot name="header">
-        <AppHeader />
+        <HeaderCatalogue v-if="route.params.catalogue" />
+        <HeaderGlobal v-else />
```
single header for consistency , maybe look at nuxt template feature to avoid huge complex app.vue file

mswertz replied:
I see this as intermediate step; we need to discuss custom landing page of UMCG versus general ones. We might need some configuration for this.

connoratrug commented on apps/nuxt3-ssr/assets/icons/Checklist.svg:
suggest making separate PR for icons, icons also look somewhat out of place with the rest of the design, but maybe we should ask de designer about that

mswertz replied:
done

connoratrug commented on apps/nuxt3-ssr/components/CohortCard.vue:
```diff
@@ -8,6 +9,7 @@ const props = withDefaults(
     cohort: ICohort;
     schema: string;
     compact?: boolean;
+    catalogue?: string;
```
name ? id ? ...

mswertz replied:
currently it is a network id. However I intend to generalize. We also call it 'schema' so I suppose we don't specify 'id' here?

connoratrug replied:
the Cohort card does not need to know about the catalogue, it only needs an link / to / url

connoratrug commented on apps/nuxt3-ssr/components/CohortDetails.vue:
```diff
@@ -1,7 +1,12 @@
 <script setup lang="ts">
+const props = defineProps({
+  cohort: { type: Object },
```
use typeScript interface

connoratrug commented on apps/nuxt3-ssr/components/CohortDetails.vue:
```diff
@@ -176,6 +186,15 @@ function collectionEventMapper(item: any) {
   };
 }
 
+function datasetMapper(item: any) {
```
any ? item struc is known from query

mswertz replied:
same holds for collectionEvent so I just copies that => would be good next step to seperate out those types so we can use them here.

connoratrug commented on apps/nuxt3-ssr/components/CohortDetails.vue:
```diff
@@ -292,6 +311,16 @@ let fundingAndAcknowledgementItems = computed(() => {
+  crumbs[`${route.params.catalogue}`] = `${route.path}/../..`;
+  crumbs["Cohorts"] = `${route.path}/..`;
+} else {
+  crumbs["Home"] = `${route.path}/../../..`;
```
silently breaks on route/struct changes , route object is a global var and suggest to be explicit about paths

mswertz replied:
fixed

connoratrug commented on apps/nuxt3-ssr/components/CohortList.vue:
moving the data fetch into the components makes it harder to keep number of fetches in check ( see rest of emx apps). keep in mind that initially the fetches are blocking the page request.

connoratrug commented on apps/nuxt3-ssr/components/LogoMobile.vue:
```diff
@@ -16,8 +17,14 @@ if (logoFileName) {
+    <img
+      v-if="image"
+      :src="image"
+      width="200"
+      height="50"
+      style="background-color: white"
+    />
```
use tailwind for themable style

mswertz replied:
bg-white?

connoratrug replied:
sure , what about images with white foreground ?

connoratrug commented on apps/nuxt3-ssr/components/NetworkCard.vue:
```diff
@@ -34,18 +35,18 @@ const iconStarClasses = computed(() => {
-    title: "Variables",
-    url: `/${props.schema}/ssr-catalogue/networks/${props.network.id}#variables`,
-  },
+  // {
```
comment out code

connoratrug commented on apps/nuxt3-ssr/components/TableRowDetails.vue:
```diff
@@ -118,23 +118,25 @@ function buildTOC(sections: ISection[]) {
+  let temp = section.meta.labels
+    ? section.meta.labels[0].value
     : section.meta.name;
+  return temp.charAt(0).toUpperCase() + temp.substr(1);
```
const + move te reusable and tested function , Case of first letter can be done in css in most cases

connoratrug commented on apps/nuxt3-ssr/components/table/Content.vue:
```diff
@@ -26,22 +27,20 @@ let orderby = {
 }
 
-fetchRows();
+await fetchRows();
```
if used in a page ( or multiple on page ) the whole thing is going to be blocking , the use of the component has no real way of knowing this

mswertz replied:
I had loading issues without the await. You believe it works without it?

connoratrug commented on apps/nuxt3-ssr/middleware/redirect-trailing-slash.global.js:
```diff
@@ -0,0 +1,8 @@
+export default defineNuxtRouteMiddleware((to, from) => {
```
why do we need this

mswertz replied:
deleted

connoratrug commented on apps/nuxt3-ssr/nuxt.config.ts:
```diff
@@ -18,7 +18,7 @@ const config = {
     // Keys within public, will be also exposed to the client-side
     public: {
-      apiBase: "http://localhost:3000/", //'https://emx2.molgeniscloud.org/',
+      apiBase: "https://data-catalogue.molgeniscloud.org/", //'https://emx2.molgeniscloud.org/',
```
keep local as default

connoratrug commented on apps/nuxt3-ssr/pages/[schema]/ssr-catalogue/[catalogue]/[resourceType]/[resource]/index.vue:
```diff
@@ -0,0 +1,3 @@
+<template>
+  <TableRowDetails />
```
suggest to look in to using the routes.ts if not using the pages feature

mswertz replied:
I think if we consider 'all' as a special filter set that we can move TableRowDetails back to be a page.

connoratrug commented on apps/nuxt3-ssr/pages/[schema]/ssr-catalogue/browse/index.vue:
the browse route feels a bit vage/undescriptive  to me, do not see a clear site hierarchy

mswertz replied:
rename to 'all' ?

connoratrug reviewed (changes requested):
preview does not work 

- global navigation confuses me 
- mixing sync ( pages stuff ) and async makes it hard to keep page load times in check
- see comments for details

mswertz commented:
> * mixing sync ( pages stuff ) and async makes it hard to keep page load times in check

fixed now?

mswertz commented:
> * components folder already has readmore component https://github.com/molgenis/molgenis-emx2/blob/master/apps/nuxt3-ssr/components/content/ReadMore.vue

I used that one right?

BrendaHijmans reviewed (commented):
functional review:
-On home page: https://preview-emx2-pr-2761.dev.molgenis.org/catalogue-demo/ssr-catalogue/, the statistics do not match what is in the tables:
*Datasets: homepage count is 51, in table view (https://preview-emx2-pr-2761.dev.molgenis.org/catalogue-demo/tables/#/Datasets) there are 87
*collected variables and harmonized variables link to the same view. Where does the number 6 and 109 come from?

-no descriptives or datasets visible in data sources details page: https://preview-emx2-pr-2761.dev.molgenis.org/catalogue-demo/ssr-catalogue/OOM/datasources/ABL

minor comments:
-I do not like the 'read more', I think this is already implied when you cut off the text and have trailing dots '...'

connoratrug commented on apps/nuxt3-ssr/components/CohortList.vue:
```diff
@@ -0,0 +1,255 @@
+const config = useRuntimeConfig();
+const pageSize = 10;
+
+useHead({ title: "Cohorts" });
```
setting globals from components is not going to work, this whole idea of moving the pages stuff into the components is breaking the modular design

mswertz replied:
I don't think I created this; originally CohortList was a page but now we render it on different places. However, if we use 'all' as a special filter then we can move this back to a page.

connoratrug commented on apps/nuxt3-ssr/components/Logo.vue:
```diff
@@ -19,7 +20,13 @@ if (logoFileName) {
+    <img
+      v-else-if="image"
+      :src="image"
+      width="200"
+      height="50"
+      class="bg-white"
```
fixed for all screen sizes

mswertz replied:
what should it be in your mind?

connoratrug commented on apps/nuxt3-ssr/components/ResourceCard.vue:
```diff
@@ -15,6 +15,9 @@ const props = withDefaults(
 );
 
+const route = useRoute();
+const catalogue = route.params.catalogue || 'all';
```
all ?

mswertz replied:
'all' is the keyword for not limiting to a specific catalogue.

mswertz commented on apps/nuxt3-ssr/components/landing/CardSecondary.vue:
```diff
@@ -1,16 +1,30 @@
   >
     <span class="mb-2 mt-2.5 block text-icon">
-      <BaseIcon :name="icon" :width="32" class="text-title" />
+      <!-- todo: the icon fill should be set via theme -->
```
do you know how?

BrendaHijmans commented:
'Browse all catalogues' does not work. "Oops!, something went wrong.
Status code: 404"

i am still missing datasets and details for data sources

BrendaHijmans commented:
Can we change "Browse OOM catalogued population and disease specific cohort studies" into something like: "Browse OOM cohort studies". Less is more. Same for "Browse OOM catalogued health and population databanks and registries". Change to: "Browse OOM databanks and registries".

connoratrug commented on apps/nuxt3-ssr/components/Button.vue:
why not `<NuxtLink><Button>Click me !</Button></NuxtLink>`

mswertz replied:
hmm, that is just my ignorance not realizing that that would have been easier indeed. I think that is indeed superior if that works because it keeps the component seperate from the nuxtlink logic.

mswertz replied:
fixed

connoratrug commented on apps/nuxt3-ssr/components/CatalogueCard.vue:
```diff
@@ -0,0 +1,131 @@
+
+const catalogue = route.params.catalogue || "all";
+
+const articleClasses = computed(() => {
+  return props.compact ? "py-5 lg:px-12.5 p-5" : "lg:px-12.5 py-12.5 px-5";
+});
+
+const subtitleClasses = computed(() => {
+  return props.compact ? "hidden md:block" : "mt-1.5 block md:inline";
+});
+
+const titleContainerClasses = computed(() => {
+  return props.compact ? "flex items-center" : "";
+});
+
+const headerClasses = computed(() => {
+  return props.compact ? "" : "items-start xl:items-center";
+});
+
+const iconStarClasses = computed(() => {
+  return props.compact ? "" : "items-baseline xl:items-center mt-0.5 xl:mt-0";
+});
```
code duplication why not use slots

mswertz replied:
can you elaborate? I would expect somehow to make components that do this for us somehow?

connoratrug replied:
something like `
<Card >
<slot header>
<slot image>
<slot description>
<slot links> 
<Card>`

  maybe with some config props. Or instead of slots a few predefined card types to choose from ( CardWithImage, CardWithSections , ...) that have props ( or combination of slots and props ) we can then wrap these with catalogue specific components <CohortCard><CardWithSections/> <CohortCard> ( in this case the cohort cards takes a number of cohort specific items and places them / passes them to the 'CardWithSections'. The CardWithSections ( typed card) or Cards with lots of optional slots can then handle all the little class setting stuff.

mswertz replied:
So I am trying to not redesign the components in this PR so I simply copied the existing pages. So do you agree if we accept the duplication for now?

I also believe that with som effort we can dramatically simplify the card creation (I would ideally have some configuration json object that is ideally an extension of emx2 so that in simple cases I just need to paste some emx2 columns, probably based on the generic cards you already made).

mswertz commented:
> It would be good to have the option to search all data resources at once, search though all data sources, databanks, cohorts.

agreed, but too much for this PR. This is just a step in the right direction.

mswertz commented:
> 'Browse all catalogues' does not work. "Oops!, something went wrong. Status code: 404"
> 
> i am still missing datasets and details for data sources

That will be a seperate PR. This one will need to change the structure to allow for the subcatalogues.

connoratrug commented:
Downloadend the branch, fixed some issue to get the build to run an look another look. I'm sorry but this can not been merged in its current state. The design completely diverges from the page design as supplied by the design company. As mentioned above it it very wordy and includes a lot of extra little labels. To me it comes across a messy and the content hierarchy is not clear to me. I suspect this is why all the texts and labels were added but this is not how web pages are used, vistors do not read all of the little sentences on a page ( people use the expectations and scan for key words ).  clicking around o bit a get completely lost ( also lots of the links are broken ).

One of most importend things a site UI needs to do is answer the question 'where am i ?". From there the visitor can decide 'where to go' ( links with keywords  ) or 'what to do' ( buttons with actions ). When a link is followed it should again be clear where the visitor is ( and where he came from and how to get back ( desoverbilty

mswertz commented:
> This PR solves a number of stakeholders requests but it's current implementation fails to answer these some of these importend ui questions. Imo the design supplied by the agency did answer these questions.

Hi Conner. This is an intermediate step so please treat it as such, and give concrete suggestions how to improve instead of complaining please. We can't afford to spend another half year staring at our belly buttons. Also, I don't agree that the agency did answer the questions in a better way, there are many places where their design is missing the mark. And that is okay, I value their help, but now we need to adapt it to the real needs and test it with our users. In the end their opinions are what matters to me, not what any design agency says. So unless you have concrete suggestions we will first merge this and take it from here.

connoratrug reviewed (changes requested):
- Fails to load font on 'home' page ( see network tab for details)
- console.log("found") , more random console.log statements
- can not view harmonisation details via grid or list
- can not view subcohort details ( page not found)
- can not view collection events ( page not found )
- cohortOnly feature flag is broken 
- error page style looks 'broken', page is means to handle error case for user, not for development debugging 
<img width="739" alt="Screenshot 2023-11-20 at 17 24 58" src="https://github.com/molgenis/molgenis-emx2/assets/47183404/76a721a8-7b6a-4b04-92b6-ddcceaac01d8">

wrong icon color on umcg theme: 
<img width="1154" alt="Screenshot 2023-11-20 at 17 37 04" src="https://github.com/molgenis/molgenis-emx2/assets/47183404/79e0648f-6f99-4130-b1dc-6ee93e6e0682">

mswertz commented:
> * Fails to load font on 'home' page ( see network tab for details)
could not reproduce, what do you mean?

> * console.log("found") , more random console.log statements
removed also from previous PRs

> * can not view harmonisation details via grid or list
> * can not view subcohort details ( page not found)
fixed

> * can not view collection events ( page not found )
fixed

> * cohortOnly feature flag is broken
this we will address in a seperate cohortsOnly PR
=> I think we should auto-redirect to '/all' OR we could have a custom landing page module here. That last thing might be easier? Also we might want to remove the 'all' breadcrumb in those cases.
(I feel us running in circles we could instead map multiple routes so it is same as before in case of cohortsOnly).

> * error page style looks 'broken', page is means to handle error case for user, not for development debugging
added debug option to only show this in dev
